File: server/src/main/java/dev/ultreon/quantum/network/system/PacketSerializer.java
package dev.ultreon.quantum.network.system;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.Serializer;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import dev.ultreon.quantum.network.PacketIO;
import dev.ultreon.quantum.network.packets.Packet;
import dev.ultreon.quantum.registry.RegistryHandle;
import dev.ultreon.quantum.server.QuantumServer;

import java.lang.reflect.InvocationTargetException;

public class PacketSerializer extends Serializer<Packet<?>> {
    private final Kryo kryo;
    private RegistryHandle registryHandle;

    public PacketSerializer(Kryo kryo, RegistryHandle registryHandle) {
        this.kryo = kryo;
        this.registryHandle = registryHandle;
    }

    @Override
    public void write(Kryo kryo, Output output, Packet<?> o) {
        PacketIO packetIO = new PacketIO(null, output.getOutputStream(), registryHandle);
        o.toBytes(packetIO);
    }

    @Override
    public Packet<?> read(Kryo kryo, Input input, Class<Packet<?>> aClass) {
        if (!Packet.class.isAssignableFrom(aClass)) throw new IllegalArgumentException("Class " + aClass.getName() + " is not a valid packet class");
        try {
            var constructor = aClass.getMethod("read", PacketIO.class);
            constructor.setAccessible(true);
            return (Packet<?>) constructor.invoke(null, new PacketIO(input.getInputStream(), null, registryHandle));
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
    }

    public Kryo getKryo() {
        return kryo;
    }
}

File: server/src/main/java/dev/ultreon/quantum/network/system/PacketSerializer.kt
package dev.ultreon.quantum.network.system

import com.esotericsoftware.kryo.Kryo
import com.esotericsoftware.kryo.Serializer
import com.esotericsoftware.kryo.io.Input
import com.esotericsoftware.kryo.io.Output
import dev.ultreon.quantum.network.PacketIO
import dev.ultreon.quantum.network.packets.Packet
import java.lang.reflect.InvocationTargetException

class PacketSerializer(val kryo: Kryo) :
  Serializer<Packet<*>>() {
  override fun write(kryo: Kryo, output: Output, o: Packet<*>) {
    val packetIO = PacketIO(null, output)
    o.toBytes(packetIO)
  }

  override fun read(kryo: Kryo, input: Input, aClass: Class<Packet<*>>): Packet<*> {
    require(Packet::class.java.isAssignableFrom(aClass)) { "Class " + aClass.name + " is not a valid packet class" }
    try {
      val constructor = aClass.getConstructor(
        PacketIO::class.java
      )
      constructor.isAccessible = true
      return constructor.newInstance(PacketIO(input, null))
    } catch (e: NoSuchMethodException) {
      throw RuntimeException(e)
    } catch (e: IllegalAccessException) {
      throw RuntimeException(e)
    } catch (e: InstantiationException) {
      throw RuntimeException(e)
    } catch (e: InvocationTargetException) {
      throw RuntimeException(e)
    }
  }
}
